highfidelity/hifi
# Experimental v3 Procedural entities with depth and normals

Edit:  hmmm something seems to be wrong with transparent procedural entities...I'll fix that when I get a chance

Test plan:
- Perform these tests on various graphics cards: NVIDIA, AMD, test on Mac
- Run this from your console:
```
Entities.addEntity({
	type: "Box",
	position: Vec3.sum(MyAvatar.position, Vec3.multiply(4.0, Quat.getForward(MyAvatar.orientation))),
	dimensions: {x: 0.5, y: 0.5, z: 0.5},
	collisionless: true,
	userData: JSON.stringify({"ProceduralEntity":{"version":3,"shaderUrl":"https://gist.githubusercontent.com/SamGondelman/932c8c03fd39b0549f0a07c0a736abd9/raw/8460af956c640d7b73c0e30ed7a4212d9876c42d/raymarch.fs","uniforms":{"scene":0}}})
})
```
- You should see some grey spheres moving around a green rippling ball that morphs into a beach ball.  They should be lit correctly by the keylight.  Try placing other lights and entities around.  The spheres should be lit by the local lights and should depth-test properly with the objects.
- Look at the balls in VR.  They should appear 3D.  Pass your hand through them.  They should intersect your hand properly (with respect to depth, not physics).
- Run the procedural entity test: https://github.com/highfidelity/hifi_tests/blob/master/tests/content/entity/procedural/test.md
- Test any other procedural content you can find (the hexagon wall in Eschatology, etc.).  They should be unchanged.
